impl/src/main/java/org/ehcache/impl/internal/store/disk/OffHeapDiskStore.java

```diff
@@ -128,21 +135,51 @@ public OffHeapDiskStore(FileBasedPersistenceContext fileBasedPersistenceContext,
   private EhcachePersistentConcurrentOffHeapClockCache<K, OffHeapValueHolder<V>> getBackingMap(long size, Serializer<K> keySerializer, Serializer<V> valueSerializer, EvictionVeto<K, OffHeapValueHolder<V>> evictionVeto) {
     File dataFile = getDataFile();
     File indexFile = getIndexFile();
+    File metadataFile = getMetadataFile();
 
-    if (dataFile.isFile() && indexFile.isFile()) {
+    if (dataFile.isFile() && indexFile.isFile() && metadataFile.isFile()) {
       try {
         return recoverBackingMap(size, keySerializer, valueSerializer, evictionVeto);
       } catch (IOException ex) {
         throw new RuntimeException(ex);
       }
     } else {
-      return createBackingMap(size, keySerializer, valueSerializer, evictionVeto);
+      try {
+        return createBackingMap(size, keySerializer, valueSerializer, evictionVeto);
+      } catch (IOException ex) {
+        throw new RuntimeException(ex);
+      }
     }
   }
 
   private EhcachePersistentConcurrentOffHeapClockCache<K, OffHeapValueHolder<V>> recoverBackingMap(long size, Serializer<K> keySerializer, Serializer<V> valueSerializer, EvictionVeto<K, OffHeapValueHolder<V>> evictionVeto) throws IOException {
     File dataFile = getDataFile();
     File indexFile = getIndexFile();
+    File metadataFile = getMetadataFile();
+
+    FileInputStream fis = new FileInputStream(metadataFile);
+    Properties properties = new Properties();
+    try {
+      properties.load(fis);
+    } finally {
+      fis.close();
+    }
+    try {
+      Class<?> persistedKeyType = classLoader.loadClass(properties.getProperty("keyType"));
+      if (!keyType.isAssignableFrom(persistedKeyType)) {
```

Comment:
That's a question I had in mind: do we allow compatible type definitions in the cache config or not? I went this route because it allows using a Cache<Object, Object> with a Store<String, String> for instance.

But thinking twice about this, I believe you're right: configured types should be exactly the same as the store types otherwise polluting the store's contents become possible.